backtrace.3
.\"	$NetBSD: backtrace.3,v 1.5 2013/08/22 17:08:43 christos Exp $
.\"
.\" Copyright (c) 2012 The NetBSD Foundation, Inc.
.\" All rights reserved.
.\"
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Christos Zoulas
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd May 26, 2012
.Dt BACKTRACE 3
.Os
.Sh NAME
.Nm backtrace
.Nd fill in the backtrace of the currently executing thread
.Sh LIBRARY
.Lb libexecinfo
.Sh SYNOPSIS
.In execinfo.h
.Ft size_t
.Fn backtrace "void **addrlist" "size_t len"
.Ft "char **"
.Fn backtrace_symbols "void * const *addrlist" "size_t len"
.Ft int
.Fn backtrace_symbols_fd "void * const *addrlist" "size_t len" "int fd"
.Ft "char **"
.Fn backtrace_symbols_fmt "void * const *addrlist" "size_t len" "const char *fmt"
.Ft int
.Fn backtrace_symbols_fmt_fd "void * const *addrlist" "size_t len" "const char *fmt" "int fd"
.Sh DESCRIPTION
The
.Fn backtrace
function places into the array pointed by
.Fa addrlist
the array of the values of the program counter for each frame called up to
.Fa len
frames.
The number of frames found (which can be fewer than
.Fa len )
is returned.
.Pp
The
.Fn backtrace_symbols_fmt
function takes an array of previously filled addresses from
.Fn backtrace
in
.Fa addrlist
of
.Fa len
elements, and uses
.Fa fmt
to format them.
The formatting characters available are:
.Bl -tag -width a -offset indent
.It Dv a
The numeric address of each element as would be printed using %p.
.It Dv n
The name of the nearest function symbol (smaller than the address element)
as determined by
.Xr dladdr 3
if the symbol was dynamic, or looked up in the executable if static and
the /proc filesystem is available to determine the executable path.
.It Dv d
The difference of the symbol address and the address element printed
using 0x%tx.
.It Dv D
The difference of the symbol addresss and the address element printed using
+0x%tx if non-zero, or nothing if zero.
.It Dv f
The filename of the symbol as determined by
.Xr dladdr 3 .
.El
.Pp
The array of formatted strings is returned as a contiguous memory address which
can be freed by a single
.Xr free 3 .
.Pp
The
.Fn backtrace_symbols
function is equivalent of calling
.Fn backtrace_symbols_fmt
with a format argument of
.Dv "%a <%n%D> at %f"
.Pp
The
.Fn backtrace_symbols_fd
and
.Fn backtrace_symbols_fmt_fd
are similar to the non _fd named functions, only instead of returning
an array or strings, they print a new-line separated array of strings in
fd, and return
.Dv 0
on success and
.Dv \-1
on failure.
.Sh RETURN VALUES
The
.Fn backtrace
function returns the number of elements that were filled in the backtrace.
The
.Fn backtrace_symbols
and
.Fn backtrace_symbols_fmt
return a string array on success, and
.Dv NULL
on failure, setting
.Va errno .
Diagnostic output may also be produced by the ELF symbol lookup functions.
.Sh SEE ALSO
.Xr dladdr 3 ,
.Xr elf 3
.Sh HISTORY
The
.Fn backtrace
library of functions first appeared in
.Nx 7.0 .
.Sh BUGS
.Bl -enum
.It
Errors should not be printed but communicated to the caller differently.
.It
Because these functions use
.Xr elf 3
this is a separate library instead of being part of libc/libutil
so that no library dependencies are introduced.
.It
The Linux versions of the functions (there are no _fmt variants) use
.Ft int
instead of
.Ft size_t
arguments.
.It
Since
.Xr dladdr 3
only deals with dynamic symbols, we need to find the symbols from the main
portion of the program.
For that we need to locate the executable, and we use procfs for
finding it, which is not portable.
.El


NEWS
This is the file NEWS for the flex package. It records user -visible
changes between releases of flex.

See the file COPYING for copying conditions.

* version 2.5.35 released 2008-02-26

** fixed bug that prevented flex from accepting certain comments in the
  scanner file (resolves bugs #1849809 and #1849805)

** fix bug that prevented headers for all functions from being generated
  (resolves bug #1628314)

** change yy_size_t to be size_t (resolves bug #1849812)

** new de, nl, pl, pt_br, vi translations from the translation project

* version 2.5.34 released 2007-12-12

** introduce yylex_init_extra; see the manual for details

** introduce %option extra-type="your_type *" (resolves bug #1744505)

** The flex program now parses multiple short concatenated options (resolves bug
  #1619820). Thanks to Petr Machata of Red Hat on this issue.

** better checking after yyalloc/yyrealloc (resolves bug #1595967)

** flex now provides for a libfl_pic.a compiled with position
   independent code. Particularly useful when including a flex scanner
   in a shared library and with more recent versions of gcc. Thanks to the Debian project for the idea.

** SourceForge feature request #1658379: Expose YY_BUF_SIZE in the
	header file.

** flex better escapes filenames with special characters in them
   (resolves bug #1623600)

** a memory leak was plugged(resolves bug #1601111)

** pattern language expanded; see the manual for details on the below
   highlights

*** pattern options added to specify patterns as case-insensitive or
    case-sensitive

*** pattern options to specify whether the "." character should match
    the newline character

*** pattern options added to allow ignoring of whitespace in patterns

*** POSIX character classes may be negated in patterns

*** patterns may now use set difference, union operators

** the manual now contains an appendix listing various common patterns
   which may be useful when writing scanners

** some memory leaks were removed from the C++ scanner (but the C++
  scanner is still experimental and may change radically without
  notice)

** c++ scanners can now use yywrap

** added new unit test for c++ and yywrap

** portability fixes to some unit tests

** flex man page and flex manual in pdf now distributed in the flex
distribution

** new ca, vi, ga, nl translations from the translation project

** flex no longer comes with an rpm spec file

** flex development now happens with automake 1.9.6

* version 2.5.33 released 2006-2-20

** all flex resources are now to be found from the website at
   http://flex.sourceforge.net/

** there was no release 2.5.32 published

** numerous bug and security fixes

** new nl, vi, sv, ro, po, ga, ca, fr, tr translations from the translation project

** upgrade to use gettext 0.12 (this now makes the "pdf" and "ps"
   targets in the build system able to be run successfully)

* version 2.5.31 released 2003-4-1

** remove --enable-maintainer-mode configure option; none of the
   Makefiles were using it and it can be unduely confusing

* version 2.5.30 released 2003-4-1

** yylineno is per-buffer in reentrant scanners

** added %top directive for placing code at the top of the generated
   scanner; see manual for details

** flex now uses m4 to generate scanners; while this means that
   scanners are more readable, it means that flex requires m4 to be
   installed; see manual for details

* version 2.5.29 released 2003-3-5

** Automatic stack management for multiple input buffers in C and C++ scanners

** moved the flex documentation to a new doc/ subdirectory

** cleanups to the yy namespace

* version 2.5.28 released 2003-2-12

** flex is now hosted at sourceforge

** Fixed trailing slash bug in YY_INPUT macro def

** Flex now warns if always-interactive is specified with fast or full

* version 2.5.27 released 2003-1-21

** flex now works with recent bison versions

** new pt_br translation from the translation project

* version 2.5.26 released 2003-1-14

** Fixed table deserialization bug on big-endian archs. Patch sent from Bryce Nichols <bryce@bnichols.org>

** yyleng has proper declarations now; this caused flex to generate
   unusable scanners for some programs

** the flex distribution now includes a spec file suitable for use
   with rpm

** some more c++ fixes

** new es translation from the translation project

** slight tweeks to the flex_int*_t types

** flex now warns about pattern ranges that might be ambiguous when
   generating a case-insensitive scanner


* version 2.5.25 released 2002-12-2

** flex now uses flex_int*_t types. For C99 systems, they are just the
   int*_t types; for non-C99 systems, we just make some typedefs

** new pt_br translation from the translation project

* version 2.5.24 released 2002-11-25

* more portability fixes

** the manual continues to be updated and edited, but it's still got a
   ways to go

** it is possible to have multiple c++ scanners in the same program again

** new turkish translation from the translation project

* version 2.5.23 released 2002-10-21

** more portability fixes

** the manual includes a title page and a table-of-contents when printed

** the test suite can be run with "make check" from the top-level
   directory

** configure now accepts the --enable-maintainer-mode option

** gettext functionality is now only available externally

** the constant FLEX_BETA is defined if flex is a beta release

** the script create-test was not included in the distribution and it
   should have been

* version 2.5.22 released 2002-10-10

** more portability fixes around how we get ahold of the integral
   types; there is a constant FLEX_NEED_INTEGRAL_TYPE_DEFINITIONS
   which you should define if you don't have the <inttypes.h> header
   file (after you complain to your C vendor for not providing a
   reasonable C environment)

** more test suite cleanups; in particular, the test suite should run
   correctly when build from a different directory


backtrace.c
/*	$NetBSD: backtrace.c,v 1.3 2013/08/29 14:58:56 christos Exp $	*/

/*-
 * Copyright (c) 2012 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Christos Zoulas.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <sys/cdefs.h>
__RCSID("$NetBSD: backtrace.c,v 1.3 2013/08/29 14:58:56 christos Exp $");

#include <sys/param.h>
#include <assert.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdint.h>
#include <stddef.h>
#include <unistd.h>
#include <fcntl.h>
#include <dlfcn.h>
#include <elf.h>

#include "execinfo.h"
#include "symtab.h"

#ifdef __linux__
#define SELF	"/proc/self/exe"
#else
#include <sys/sysctl.h>
#define SELF	"/proc/curproc/file"
#endif

static int
open_self(int flags)
{
	const char *pathname = SELF;
#ifdef KERN_PROC_PATHNAME
	static const int name[] = {
		CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1,
	};
	char path[MAXPATHLEN];
	size_t len;

	len = sizeof(path);
	if (sysctl(name, __arraycount(name), path, &len, NULL, 0) != -1)
		pathname = path;
#endif
	return open(pathname, flags);
}


static int __printflike(4, 5)
rasprintf(char **buf, size_t *bufsiz, size_t offs, const char *fmt, ...)
{
	for (;;) {
		size_t nbufsiz;
		char *nbuf;

		if (*buf && offs < *bufsiz) {
			va_list ap;
			int len;

			va_start(ap, fmt);
			len = vsnprintf(*buf + offs, *bufsiz - offs, fmt, ap);
			va_end(ap);

			if (len < 0 || (size_t)len < *bufsiz - offs)
				return len;
			nbufsiz = MAX(*bufsiz + 512, (size_t)len + 1);
		} else
			nbufsiz = MAX(offs, *bufsiz) + 512;
			
		nbuf = realloc(*buf, nbufsiz);
		if (nbuf == NULL)
			return -1;
		*buf = nbuf;
		*bufsiz = nbufsiz;
	}
}

/*
 * format specifiers:
 *	%a	= address
 *	%n	= symbol_name
 *	%d	= symbol_address - address
 *	%D	= if symbol_address == address "" else +%d
 *	%f	= filename
 */
static ssize_t
format_string(char **buf, size_t *bufsiz, size_t offs, const char *fmt,
    Dl_info *dli, const void *addr)
{
	ptrdiff_t diff = (const char *)addr - (const char *)dli->dli_saddr;
	size_t o = offs;
	int len;

	for (; *fmt; fmt++) {
		if (*fmt != '%')
			goto printone;
		switch (*++fmt) {
		case 'a':
			len = rasprintf(buf, bufsiz, o, "%p", addr);
			break;
		case 'n':
			len = rasprintf(buf, bufsiz, o, "%s", dli->dli_sname);
			break;
		case 'D':
			if (diff)
				len = rasprintf(buf, bufsiz, o, "+0x%tx", diff);
			else
				len = 0;
			break;
		case 'd':
			len = rasprintf(buf, bufsiz, o, "0x%tx", diff);
			break;
		case 'f':
			len = rasprintf(buf, bufsiz, o, "%s", dli->dli_fname);
			break;
		default:
		printone:
			len = rasprintf(buf, bufsiz, o, "%c", *fmt);
			break;
		}
		if (len == -1)
			return -1;
		o += len;
	}
	return o - offs;
}

static ssize_t
format_address(symtab_t *st, char **buf, size_t *bufsiz, size_t offs,
    const char *fmt, const void *addr)
{
	Dl_info dli;

	memset(&dli, 0, sizeof(dli));
	(void)dladdr(addr, &dli);
	if (st)
		symtab_find(st, addr, &dli);

	if (dli.dli_sname == NULL)
		dli.dli_sname = "???";
	if (dli.dli_fname == NULL)
		dli.dli_fname = "???";
	if (dli.dli_saddr == NULL)
		dli.dli_saddr = (void *)(intptr_t)addr;

	return format_string(buf, bufsiz, offs, fmt, &dli, addr);
}

char **
backtrace_symbols_fmt(void *const *trace, size_t len, const char *fmt)
{

	static const size_t slen = sizeof(char *) + 64;	/* estimate */
	char *ptr;
	symtab_t *st;
	int fd;

	if ((fd = open_self(O_RDONLY)) != -1)
		st = symtab_create(fd, -1, STT_FUNC);
	else
		st = NULL;

	if ((ptr = calloc(len, slen)) == NULL)
		goto out;

	size_t psize = len * slen;
	size_t offs = len * sizeof(char *);

	/* We store only offsets in the first pass because of realloc */
	for (size_t i = 0; i < len; i++) {
		ssize_t x;
		((char **)(void *)ptr)[i] = (void *)offs;
		x = format_address(st, &ptr, &psize, offs, fmt, trace[i]);
		if (x == -1) {
			free(ptr);
			ptr = NULL;
			goto out;
		}
		offs += x;
		ptr[offs++] = '\0';
		assert(offs < psize);
	}

	/* Change offsets to pointers */
	for (size_t j = 0; j < len; j++)
		((char **)(void *)ptr)[j] += (intptr_t)ptr;

out:
	symtab_destroy(st);
	if (fd != -1)
		(void)close(fd);

	return (void *)ptr;
}

int
backtrace_symbols_fd_fmt(void *const *trace, size_t len, int fd,
    const char *fmt)
{
	char **s = backtrace_symbols_fmt(trace, len, fmt);
	if (s == NULL)
		return -1;
	for (size_t i = 0; i < len; i++)
		if (dprintf(fd, "%s\n", s[i]) < 0)
			break;
	free(s);
	return 0;
}

static const char fmt[] = "%a <%n%D> at %f";

char **
backtrace_symbols(void *const *trace, size_t len)
{
	return backtrace_symbols_fmt(trace, len, fmt);
}

int
backtrace_symbols_fd(void *const *trace, size_t len, int fd)
{
	return backtrace_symbols_fd_fmt(trace, len, fd, fmt);
}
